Mixpanel/MixpanelExceptionHandler.m

```diff
@@ -93,88 +108,66 @@ - (void)addMixpanelInstance:(Mixpanel *)instance {
-    if (exceptionCount > UncaughtExceptionMaximum)
-    {
-        return;
+    if (exceptionCount <= UncaughtExceptionMaximum) {
+        NSArray *callStack = [MixpanelExceptionHandler backtrace];
+        NSMutableDictionary *userInfo = [[exception userInfo] mutableCopy];
+        userInfo[UncaughtExceptionHandlerAddressesKey] = callStack;
+        NSException *exceptionWithCallstack = [NSException exceptionWithName:[exception name]
+                                                                      reason:[exception reason]
+                                                                    userInfo:userInfo];
+
+        [handler mp_handleUncaughtException:exceptionWithCallstack];
     }
 
-    NSArray *callStack = [MixpanelExceptionHandler backtrace];
-    NSMutableDictionary *userInfo =
-    [NSMutableDictionary dictionaryWithDictionary:[exception userInfo]];
-    [userInfo
-     setObject:callStack
-     forKey:UncaughtExceptionHandlerAddressesKey];
-
-    [MixpanelExceptionHandler performSelectorOnMainThread:@selector(mp_handleUncaughtException:)
-                              withObject:[NSException
-                                          exceptionWithName:[exception name]
-                                          reason:[exception reason]
-                                          userInfo:userInfo] waitUntilDone:YES];
+    if (handler.defaultExceptionHandler) {
+        handler.defaultExceptionHandler(exception);
+    }
 }
 
 
-+ (void) mp_handleUncaughtException:(NSException *)exception {
-    MixpanelExceptionHandler *handler = [MixpanelExceptionHandler sharedHandler];
-
+- (void) mp_handleUncaughtException:(NSException *)exception {
     // Archive the values for each Mixpanel instance
-    for (Mixpanel *instance in handler.mixpanelInstances) {
-        [instance archive];
+    for (Mixpanel *instance in self.mixpanelInstances) {
         NSMutableDictionary *properties = [[NSMutableDictionary alloc] init];
         [properties setValue:[exception reason] forKey:@"$ae_crashed_reason"];
         [instance track:@"$ae_crashed" properties:properties];
-        dispatch_sync(instance.serialQueue, ^{});
+        dispatch_sync(instance.serialQueue, ^{
+            [instance archive];
+        });
     }
-
-    NSSetUncaughtExceptionHandler(NULL);
-    signal(SIGABRT, SIG_DFL);
-    signal(SIGILL, SIG_DFL);
-    signal(SIGSEGV, SIG_DFL);
-    signal(SIGFPE, SIG_DFL);
-    signal(SIGBUS, SIG_DFL);
-    signal(SIGPIPE, SIG_DFL);
     NSLog(@"Encountered an uncaught exception. All Mixpanel instances were archived.");
-
     NSLog(@"%@", [NSString stringWithFormat:@"Debug details follow:\n%@\n%@",
                   [exception reason],
                   [[exception userInfo] objectForKey:UncaughtExceptionHandlerAddressesKey]]);
-    if (handler.defaultExceptionHandler) {
-        // Ensure the existing handler gets called once we're finished
-        handler.defaultExceptionHandler(exception);
```

Comment:
don't we want to call the default handler for the created exception?

Reply:
i moved it to the exception handler/signal handler functions https://github.com/mixpanel/mixpanel-iphone/pull/718/files#diff-5d925142f5bb1fb4b058808a4fcdeb10R152